plugins/inputs/cloudwatch/cloudwatch.go

```diff
@@ -311,89 +396,192 @@ func (c *CloudWatch) fetchNamespaceMetrics() ([]*cloudwatch.Metric, error) {
 /*
  * Gather given Metric and emit any error
  */
-func (c *CloudWatch) gatherMetric(
-	acc telegraf.Accumulator,
-	metric *cloudwatch.Metric,
-) error {
-	params := c.getStatisticsInput(metric)
-	resp, err := c.client.GetMetricStatistics(params)
+func (c *CloudWatch) gatherMetrics(
+	params *cloudwatch.GetMetricDataInput,
+) ([]*cloudwatch.MetricDataResult, error) {
+	resp, err := c.client.GetMetricData(params)
 	if err != nil {
-		return err
+		return nil, errors.New("Failed to get metric data - " + err.Error())
 	}
 
-	for _, point := range resp.Datapoints {
+	// todo: determine if we need to handle pagination.
+	if resp.NextToken != nil {
+		log.Println("[inputs.cloudwatch] W! UNHANDLED PAGINATED RESULTS!!", *resp.NextToken)
```

Comment:
when I've written AWS code I typically put it in a for {} loop and break when NextToken is nil.

@gianarb how are you doing AWS paging code now-a-days?  Do you just do a forever loop?

Reply:
Also how often do you get paginated results, so far neither I nor those testing these changes have come across paginated results.